src/puppetlabs/puppetdb/cli/shovel.clj

```diff
@@ -0,0 +1,378 @@
+          (finally
+            (.acknowledge message)))))))
+
+(defn consume-everything
+  "Attempts to read all available messages from ActiveMQ, giving up
+  after 5 seconds if one is not available"
+  [^MessageConsumer consumer process-message]
+  (try
+    (loop [id 0]
+      ;; We'll wait 5 seconds for a message otherwise we'll assume the Queue is
+      ;; drained
+      (when-let [msg (.receive consumer 5000)]
+        (process-message id msg)
+        (recur (inc id))))
+    (catch javax.jms.IllegalStateException e
+      (log/info (i18n/trs "Received IllegalStateException, shutting down")))
+    (catch Exception e
+      (log/error e))))
+
+(defn create-mq-receiver
+  [^ConnectionFactory conn-pool endpoint process-message]
+  (with-open [connection (create-connection conn-pool)
+              session (create-session connection)
+              consumer (.createConsumer session (create-queue session endpoint))]
+    (log/info (i18n/trs "Processing messages from the Queue: {0}" endpoint))
+    (consume-everything consumer process-message)))
+
+(defn create-scheduler-receiver
+  "Drains the scheduler queue (where PuppetDB commands are retried)
+  and enqueues them into the normal queue"
+  [^ConnectionFactory conn-pool endpoint process-message]
+  (with-open [connection (create-connection conn-pool)
+              session (create-session connection)]
+    (let [request-browse (.createTopic session ScheduledMessage/AMQ_SCHEDULER_MANAGEMENT_DESTINATION)
+          browse-dest (.createTemporaryQueue session)]
+      (with-open [producer (.createProducer session request-browse)
+                  consumer (.createConsumer session browse-dest)]
+        (let [request (doto (.createMessage session)
+                        (.setStringProperty ScheduledMessage/AMQ_SCHEDULER_ACTION
+                                            ScheduledMessage/AMQ_SCHEDULER_ACTION_BROWSE)
+                        (.setJMSReplyTo browse-dest))
+              process-message-and-remove (fn [^Message msg]
+                                           (try
+                                             (process-message msg)
+                                             (let [msg-id (.getStringProperty msg
+                                                                              ScheduledMessage/AMQ_SCHEDULED_ID)
+                                                   remove (doto (.createMessage session)
+                                                            (.setStringProperty ScheduledMessage/AMQ_SCHEDULER_ACTION
+                                                                                ScheduledMessage/AMQ_SCHEDULER_ACTION_REMOVE)
+                                                            (.setStringProperty ScheduledMessage/AMQ_SCHEDULED_ID msg-id))]
+                                               (.send producer remove))
+                                             (catch Exception e nil)))]
```

Comment:
Is this just best-effort, or wondered if it might retry, or maybe this could/should discard?